File: kc.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
# coding=utf-8

import argparse
import os
import json
import hashlib
import datetime
import struct
import unicodedata


class Sectionizer:
    def __init__(self, fn, perm):
        self.f = open(fn, perm)
        header = self.f.read(78)
        self.ident = header[0x3C:0x3C+8]
        if self.ident != b'BOOKMOBI':
            raise ValueError('invalid file format')
        num_sections, = struct.unpack_from('>H', header, 76)
        sections = self.f.read(num_sections*8)
        self.sections = struct.unpack_from('>%dL' % (num_sections*2), sections, 0)[::2] + (0xfffffff, )

    def load_section(self, section):
        before, after = self.sections[section:section+2]
        self.f.seek(before)
        return self.f.read(after - before)


def get_mobi_asin(file_path):
    try:
        sections = Sectionizer(file_path, 'rb')
        header = sections.load_section(0)
        len_mobi = struct.unpack_from('>L', header, 20)[0] + 16
        len_exth, = struct.unpack_from('>L', header, len_mobi + 4)
        exth_records = header[len_mobi:len_mobi + len_exth][12:]
        exth = dict()
        while len(exth_records) > 8:
            rectype, reclen = struct.unpack_from('>LL', exth_records)
            recdata = exth_records[8:reclen]
            exth[rectype] = recdata
            exth_records = exth_records[reclen:]
        if 113 in exth:
            return exth[113].decode("utf-8")
        return ""
    except BaseException as e:
        print(e)
        return ""


parser = argparse.ArgumentParser()
parser.add_argument('--kindleDir', default='/Volumes/Kindle')
rootDir = parser.parse_args().kindleDir.rstrip('/')
documentsDir = rootDir + '/documents/'
collectionsPath = rootDir + "/system/collections.json"

with open(collectionsPath, encoding='utf8') as json_data:
    oldCollections = json.load(json_data)

dirs = [name for name in os.listdir(documentsDir) if os.path.isdir(os.path.join(documentsDir, name))]

newCollections = {}
for d in dirs:
    colName = d + '@en-US'
    newCollections[colName] = {'lastAccess': 0, 'items': []}
    files = [name for name in os.listdir(documentsDir+'/'+d) if os.path.isfile(os.path.join(documentsDir+'/'+d, name))]
    for f in files:
        path = unicodedata.normalize('NFC', '/mnt/us/documents/' + d + '/' + f)
        filename, file_extension = os.path.splitext(f)
        if file_extension.lower() == '.mobi':
            asin = get_mobi_asin(documentsDir + d + '/' + f)
            if asin == "":
                newCollections[colName]['items'].append('*' + hashlib.sha1(path.encode('utf-8')).hexdigest())
            else:
                newCollections[colName]['items'].append('#' + get_mobi_asin(documentsDir + d + '/' + f) + '^EBOK')
        elif file_extension.lower() == '.pdf':
            newCollections[colName]['items'].append('*' + hashlib.sha1(path.encode('utf-8')).hexdigest())
        else:
            print('No book ext: ' + f)

os.rename(collectionsPath, collectionsPath + '.BU.' + datetime.datetime.now().strftime("%d-%m-%y-%H:%M:%S"))
with open(collectionsPath, 'w') as outfile:
    json.dump(newCollections, outfile)

print('Done!')

File: kc.php
#!/usr/bin/env php
<?php

$rootDir = rtrim($argv[1] ?? '/Volumes/Kindle', '/');
$dirs = glob("{$rootDir}/documents/*", GLOB_ONLYDIR);
$collectionsPath = "{$rootDir}/system/collections.json";
//$oldCollections = json_decode(file_get_contents($collectionsPath), true);


echo sha1(Normalizer::normalize('/mnt/us/documents/math/Грокаем Алгоритмы. Иллюстрированное пособие для программистов и любопытствущих - Бхаргава А. - 2017.PDF'));
die();

$newCollections = [];
foreach ($dirs as $dir) {
    $colName = basename($dir) . '@en-US';
    $newCollections[$colName]['lastAccess'] = 0;
    $filePaths = array_diff(glob("$dir/*"), glob("$dir/*", GLOB_ONLYDIR));
    foreach ($filePaths as $filePath) {
        $ext = strtolower(pathinfo($filePath, PATHINFO_EXTENSION));
        if (! in_array($ext, ['pdf','mobi'])) {
            echo "No book ext: \"{$filePath}\"\r\n";
            continue;
        }
        $path = '/mnt/us' . substr($filePath, strlen($rootDir));
        $newCollections[$colName]['items'][] = '*' . sha1(Normalizer::normalize($path));
    }
}

$stat = [];

$colNames = array_unique(array_merge(array_keys($newCollections), array_keys($oldCollections)));
foreach ($colNames as $colName) {
    $newCol = $newCollections[$colName] ?? [];
    $newCount = count($newCol['items'] ?? []);
    $oldCol = $oldCollections[$colName] ?? [];
    $oldCount = count($oldCol['items'] ?? []);
    $stat[] = ['name' => substr($colName, 0, -6), 'new' => $newCount, 'old' => $oldCount];
}
usort($stat, function ($a, $b) {
    return abs($a['new']-$a['old']) < abs($b['new']-$b['old']);
});
echo "\r\nStat. New/Old Collection: \r\n";
foreach ($stat as $line) {
    echo "{$line['name']}: {$line['new']}/{$line['old']}" . "\r\n";
}

$ok = rename($collectionsPath, $collectionsPath . '.' . date("Y-m-d-H:i:s"));
if (!$ok) {
    echo "Error while making backup.";
    return;
}
$ok = file_put_contents($collectionsPath, json_encode($newCollections));
if ($ok) {
    echo "\r\nDone!";
}

